core/src/main/java/com/linecorp/armeria/client/Endpoint.java

```diff
@@ -537,6 +553,50 @@ public String authority() {
         return authority;
     }
 
+    /**
+     * Returns the attribute value associated with the given {@link AttributeKey} of this endpoint, or
+     * {@code null} if there's no value set by {@link #setAttr(AttributeKey, Object)} or
+     * {@link #setAttrs(Iterable)}.
+     */
+    @Nullable
+    public <T> T attr(AttributeKey<T> key) {
+        @Nullable
+        final AttributeMap attributeMap = this.attributes;
+        if (attributeMap != null) {
+            return attributeMap.attr(key).get();
+        }
+        return null;
+    }
+
+    /**
+     * Associates the specified value with the given {@link AttributeKey} in this endpoint.
+     *
+     * @return the old value that has been replaced.
+     *
+     */
+    @Nullable
+    public <T> T setAttr(AttributeKey<T> key, @Nullable T value) {
+        return attributes().attr(key).getAndSet(value);
+    }
+
+    /**
+     * Associates a collection of {@link AttributeKey} with value to this endpoint.
+     */
+    public <T> void setAttrs(Iterable<? extends Entry<AttributeKey<T>, T>> attributes) {
+        final AttributeMap attributeMap = attributes();
+        attributes.forEach(entry -> attributeMap.attr(entry.getKey()).set(entry.getValue()));
+    }
+
+    private AttributeMap attributes() {
+        @Nullable
+        final AttributeMap attributeMap = this.attributes;
+        if (attributeMap != null) {
+            return attributeMap;
+        }
+        attributesUpdater.compareAndSet(this, null, new DefaultAttributeMap());
+        return requireNonNull(this.attributes);
```

Comment:
How about:
```suggestion
        final DefaultAttributeMap defaultAttributeMap = new DefaultAttributeMap();
        if (attributesUpdater.compareAndSet(this, null, defaultAttributeMap)) {
            return defaultAttributeMap;
        } else {
            @Nullable
            final AttributeMap attributes = this.attributes;
            assert attributes != null;
            return attributes;
        }
```
so that we don't access the volatile field twice?